scripts/lint_jjb.py

```diff
@@ -89,6 +96,67 @@ def parse_file_name(in_dir, in_file):
     return 0
 
 
+# Ensure all jobs have a retention policy
+# Structure:
+#   job:
+#     properties:
+#      - build-discarder:
+#          days-to-keep: 3
+
+def check_retention_value(job, value_name, vmin, vmax):
+    value = jmespath.search(
+        'properties[*]."build-discarder"."{vn}"'.format(vn=value_name), job)[0]
+
+    # try and convert to int, if that fails, check for templated value.
+    try:
+        days = int(value)
+    except ValueError as e:
+        # Can't check templated values, assume they are ok.
+        if '{' in value and '}' in value:
+            return
+        else:
+            raise e
+
+    if days <= vmin or days > vmax:
+        raise ValueError("Value for {vn} out of range. "
+                         "Value: {v}, min: {min}, max: {max}"
+                         .format(vn=value_name, v=value,
+                                 min=vmin, max=vmax))
+
+
+def check_retention(job, in_file):
+    error_message = ("{f}/{j}: Valid build-discarder config not found."
+                     " Please ensure days-to-keep (0 < x <= 365) or"
+                     " num-to-keep (0 < x <= 100) is specified."
+                     " Properties: {p}\n"
+                     .format(j=job['name'], f=in_file,
+                             p=job.get('properties')))
+
+    error_counter = 0
+
+    try:
+        check_retention_value(job, 'days-to-keep', 0, 365)
+    except (TypeError, IndexError):
+        error_counter += 1
+    except ValueError as e:
+        print e
+        error_counter += 1
+
+    try:
+        check_retention_value(job, 'num-to-keep', 0, 100)
+    except (TypeError, IndexError):
+        error_counter += 1
+    except ValueError as e:
+        print e
+        error_counter += 1
+
+    if (error_counter > 1):
+        sys.stderr.write(error_message)
+        return 1
+
+    return 0
+
+
```

Comment:
`out` defined on line 175 (I can't comment on that part of the file) should be updated so that it doesn't mislead on failure.

Reply:
Done, also updated readme with retention policy requirement.